File: rdos-validation-tools/rdo-validation-tool-java/src/main/java/com/rms/rdo/validation/datatype/Validator.java
package com.rms.rdo.validation.datatype;

import java.text.SimpleDateFormat;
import java.util.List;
import java.util.Optional;
import java.util.function.Supplier;
import java.util.stream.Collectors;

public class Validator {

    public static <T> Optional<String> validateNumericType(Supplier<T> value, String dataType) {
        try {
            value.get();
            return Optional.empty();
        } catch (Exception ex) {
            return Optional.of("The value is not a " + dataType);
        }
    }

    public static Optional<String> validateDateAndTimeDataType(String value, List<String> formats, String dataType) {
        List<String> result = formats.stream().map(dateFormat -> isDateValid(value, dateFormat, dataType))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.toList());
        if (result.size() == formats.size()) {
            return Optional.of(result.get(0));
        } else {
            return Optional.empty();
        }
    }

    public static <T> Optional<String> validateBitType(String value, String dataType) {
        if (value.equals("0") || value.equals("1")) {
            return Optional.empty();
        } else {
            return Optional.of("The value is not a " + dataType);
        }
    }

    private static Optional<String> isDateValid(String value, String formats, String dataType) {
        try {
            SimpleDateFormat formatter = new SimpleDateFormat(formats);
            formatter.parse(value);
            return Optional.empty();
        } catch (Exception ex) {
            return Optional.of("The value is not a " + dataType);
        }
    }
}

File: rdos-validation-tools/rdo-validation-tool-scala/src/main/scala/com/rms/rdo/validation/datatype/Validator.scala
package com.rms.rdo.validation.datatype

import java.text.SimpleDateFormat

object Validator {

  def validateNumericType[T](value: Option[T], dataType: String): Option[String] = {
    value match {
      case Some(_) => None
      case None => Some(s"The value is not a $dataType")
    }
  }

  def validateDateAndTimeDataType(value: String, formats: List[String], dataType: String): Option[String] = {
    val result = formats.flatMap(dateFormat => isDateValid(value, dateFormat, dataType))
    if (result.length == formats.length) result.headOption else None
  }

  def validateBitType(value: String, dataType: String): Option[String] = {
    if (value == "0" || value == "1") None else Some(s"The value is not a $dataType")
  }

  private def isDateValid(value: String, formats: String, dataType: String): Option[String] = {
    try {
      val formatter = new SimpleDateFormat(formats)
      formatter.parse(value)
      None
    } catch {
      case _: Exception => Some(s"The value is not a $dataType")
    }
  }

}
